File: app/src/main/java/com/example/paritosh/stormy/WeatherDataProvider.java
package com.example.paritosh.stormy;

import android.support.annotation.NonNull;

import com.example.paritosh.stormy.model.CurrentWeather;
import com.example.paritosh.stormy.model.HourlyForecastModel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class WeatherDataProvider {

    private String apiKey = "cde93abb7a85053efe6dcd43de69e1fa";
    private double longitude = 42.3601;
    private double latitude = -71.0589;

    private String forecastURL = "https://api.darksky.net/forecast/"
            + apiKey + "/" + longitude + "," + latitude;

    public void getCurrentWeather(final OnWeatherApiResponse onWeatherApiResponse) {

        OkHttpClient client = new OkHttpClient();

        // Builder Pattern
        Request request = new Request.Builder()
                .url(forecastURL)
                .build();

        Call call = client.newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                onWeatherApiResponse.onError(e);
            }

            @Override
            public void onResponse(@NonNull Call call, Response response) throws IOException {

                String jsonData = response.body().string();

                try {
                    onWeatherApiResponse.onSuccess(parseCurrentWeather(jsonData));
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        });

    }

    private CurrentWeather parseCurrentWeather(String jsonData) throws JSONException {

        JSONObject jsonObject = new JSONObject(jsonData);
        CurrentWeather currentWeather = new CurrentWeather();

        currentWeather.setLocationLabel(jsonObject.getString("timezone"));

        JSONObject currently = jsonObject.getJSONObject("currently");

        currentWeather.setTime(currently.getLong("time"));
        currentWeather.setIcon(currently.getString("icon"));
        currentWeather.setTemperature(currently.getDouble("temperature"));
        currentWeather.setHumidity(currently.getDouble("humidity"));
        currentWeather.setPrecipChance(currently.getDouble("precipProbability"));
        currentWeather.setSummary(currently.getString("summary"));

        ArrayList<HourlyForecastModel> hourly = new ArrayList<>();

        JSONObject hourlyObject = jsonObject.getJSONObject("hourly");
        JSONArray hourlyData = hourlyObject.getJSONArray("data");
        for (int i = 0; i < hourlyData.length(); i++) {
            HourlyForecastModel hour = new HourlyForecastModel();
            JSONObject data = hourlyData.getJSONObject(i);

            hour.setSummary(data.getString("summary"));
            hour.setTemperature(data.getDouble("temperature"));
            hour.setIcon(Utils.getIconId(data.getString("icon")));
            hour.setTime(Utils.getReadableTime(data.getLong("time"), currentWeather.getLocationLabel()));

            hourly.add(i, hour);
        }
        currentWeather.setHourly(hourly);
        return currentWeather;

    }

    public interface OnWeatherApiResponse {
        void onSuccess(CurrentWeather weather);

        void onError(Throwable t);
    }
}

File: app/src/main/java/com/example/paritosh/stormy/WeatherDataProvider.kt
package com.example.paritosh.stormy

import com.example.paritosh.stormy.model.CurrentWeather
import com.example.paritosh.stormy.model.HourlyForecastModel
import okhttp3.*
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.util.*

class WeatherDataProvider {

    private val apiKey: String = "cde93abb7a85053efe6dcd43de69e1fa"
    private val longitude = 42.3601
    private val latitude = -71.0589
    private val forecastURL = "https://api.darksky.net/forecast/$apiKey/$longitude,$latitude"

    fun getCurrentWeather(onWeatherApiResponse: OnWeatherApiResponse) {

        val client = OkHttpClient()

        val request = Request.Builder()
                .url(forecastURL)
                .build()

        val call = client.newCall(request)
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                onWeatherApiResponse.onError(e)
            }

            override fun onResponse(call: Call, response: Response) {


                val jsonData = response.body()!!.string()
                try {
                    onWeatherApiResponse.onSuccess(parseCurrentWeather(jsonData))
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        })
    }

    private fun parseCurrentWeather(jsonData: String): CurrentWeather {

        val jsonObject = JSONObject(jsonData)
        val currently = jsonObject.getJSONObject("currently")

        val hourly = ArrayList<HourlyForecastModel>()
        val hourlyObject = jsonObject.getJSONObject("hourly")
        val hourlyData = hourlyObject.getJSONArray("data")

        for (i in 0 until hourlyData.length()) {

            val data = hourlyData.getJSONObject(i)
            val hourlyForecastModel = HourlyForecastModel()

            hourlyForecastModel.summary = data.getString("summary")
            hourlyForecastModel.temperature = data.getDouble("temperature")
            hourlyForecastModel.icon = Utils.getIconId(data.getString("icon"))
            hourlyForecastModel.time = Utils.getReadableTime(
                    data.getLong("time"),
                    jsonObject.getString("timezone")
            )

            hourly.add(i, hourlyForecastModel)
        }


        return CurrentWeather(
                jsonObject.getString("timezone"), currently.getString("time"), currently.getLong("time"),
                currently.getDouble("temperature"), currently.getDouble("humidity"),
                currently.getDouble("precipProbability"), currently.getString("summary"),
                hourly
        )
    }

    interface OnWeatherApiResponse {
        fun onSuccess(currentWeather: CurrentWeather)

        fun onError(t: Throwable)
    }
}